ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -98,86 +102,109 @@ func (r *RayServiceReconciler) Reconcile(ctx context.Context, request ctrl.Reque
+		if ctrlResult, isHealthy, err = r.reconcileServe(ctx, rayServiceInstance, pendingRayClusterInstance, false); err != nil {
+			return ctrlResult, err
 		}
+	} else {
+		rayServiceInstance.Status.ActiveServiceStatus = rayv1alpha1.RayServiceStatus{}
+		rayServiceInstance.Status.PendingServiceStatus = rayv1alpha1.RayServiceStatus{}
 	}
 
-	var isHealthy bool
-	if isHealthy, err = r.getAndCheckServeStatus(rayServiceInstance, rayDashboardClient); err != nil {
-		err = r.updateState(ctx, rayServiceInstance, rayv1alpha1.FailGetServeDeploymentStatus, err)
-		return ctrl.Result{}, err
+	if !isHealthy {
+		return ctrl.Result{RequeueAfter: ServiceRestartRequeueDuration}, nil
 	}
 
-	log.Info("Check serve health", "isHealthy", isHealthy)
+	rayClusterInstance := activeRayClusterInstance // RayCluster instance which needs reconcile.
+	if pendingRayClusterInstance != nil {
+		rayClusterInstance = pendingRayClusterInstance
+	}
 
-	if isHealthy {
-		rayServiceInstance.Status.ServiceStatus = rayv1alpha1.Running
-		if err := r.Status().Update(ctx, rayServiceInstance); err != nil {
-			return ctrl.Result{}, err
-		}
-	} else {
-		rayServiceInstance.Status.RayClusterName = ""
-		rayServiceInstance.Status.ServiceStatus = rayv1alpha1.Restarting
-		if err := r.Status().Update(ctx, rayServiceInstance); err != nil {
+	if rayClusterInstance != nil {
+		if err := r.reconcileIngress(ctx, rayServiceInstance, rayClusterInstance); err != nil {
+			err = r.updateState(ctx, rayServiceInstance, rayv1alpha1.FailedToUpdateIngress, err)
 			return ctrl.Result{}, err
 		}
-
-		r.ServeDeploymentConfigs.Remove(request.NamespacedName.String())
-
-		// restart raycluster
-		if err := r.Delete(ctx, rayClusterInstance); err != nil {
-			err = r.updateState(ctx, rayServiceInstance, rayv1alpha1.FailDeleteRayCluster, err)
+		if err := r.reconcileServices(ctx, rayServiceInstance, rayClusterInstance); err != nil {
+			err = r.updateState(ctx, rayServiceInstance, rayv1alpha1.FailedToUpdateService, err)
 			return ctrl.Result{}, err
 		}
+	}
 
-		log.V(1).Info("Deleted rayCluster for rayService run", "rayCluster", rayClusterInstance)
-		// Wait a while for the cluster delete
-		return ctrl.Result{RequeueAfter: RayServiceRestartRequeueDuration}, nil
+	if errStatus := r.Status().Update(ctx, rayServiceInstance); errStatus != nil {
+		rayServiceLog.Error(err, "Fail to update status of RayService", "rayServiceInstance", rayServiceInstance)
+		return ctrl.Result{}, err
 	}
 
-	return ctrl.Result{RequeueAfter: RayServiceDefaultRequeueDuration}, nil
+	return ctrl.Result{RequeueAfter: ServiceDefaultRequeueDuration}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&rayv1alpha1.RayService{}).
+		Owns(&rayv1alpha1.RayCluster{}).
+		Owns(&corev1.Service{}).
+		Owns(&networkingv1.Ingress{}).
```

Comment:
does this stream service and ingress update into controller?

Reply:
also what about the configmap?

Reply:
We do not have configmap. If service or ingress updates, the controller will know

Reply:
I think the idea is that we want to detect, say, if someone deletes the RayService's service.
There's no config map :)